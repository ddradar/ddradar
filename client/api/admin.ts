import type { NuxtHTTPInstance } from '@nuxt/http'

import type { SongInfo } from '~/api/song'
import type { NotificationSchema } from '~/core/db/notification'

type NotificationRequest = Omit<
  NotificationSchema,
  'id' | 'sender' | 'timeStamp'
> & {
  id?: string
  timeStamp?: number
}

/**
 * Call "Post Notification" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postNotification
 */
export function postNotification(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: NotificationRequest
) {
  return $http.$post<NotificationRequest>('/api/v1/admin/notification', {
    ...body,
    sender: 'SYSTEM',
  })
}

/**
 * Call "Post Song Information" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/postSongInfo
 */
export function postSongInfo(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: SongInfo
) {
  return $http.$post<SongInfo>('/api/v1/admin/songs', body)
}
