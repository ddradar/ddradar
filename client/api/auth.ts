import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

type AuthResult = {
  clientPrincipal: Api.ClientPrincipal | null
}

/**
 * Call "Get Credentials" API.
 * @see https://docs.microsoft.com/azure/static-web-apps/user-information
 */
export async function getClientPrincipal(
  $http: Pick<NuxtHTTPInstance, '$get'>
) {
  const { clientPrincipal } = await $http.$get<AuthResult>('/.auth/me')
  return clientPrincipal
}
