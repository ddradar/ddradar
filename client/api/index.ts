export const apiPrefix = '/api/v1'
