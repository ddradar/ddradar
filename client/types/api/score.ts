/**
 * Object type returned by `/api/v1/scores/{:songId}/{:playStyle}/{:difficulty}`
 * @see https://github.com/ddradar/ddradar/blob/master/api/getChartScore/README.md
 */
export type UserScore = {
  userId: string
  userName: string
  /** Song id that depend on official site. ^([01689bdiloqDIOPQ]*){32}$ */
  songId: string
  songName: string
  playStyle: 1 | 2
  difficulty: 0 | 1 | 2 | 3 | 4
  /** Normal score */
  score: number
  exScore?: number
  maxCombo?: number
  /** 0: Failed, 1: Assisted Clear 2: Clear, 3: LIFE4, 4: Good FC (Full Combo), 5: Great FC, 6: PFC, 7: MFC */
  clearLamp: ClearLamp
  /** Clear rank (E-AAA) */
  rank: string
}

/** 0: Failed, 1: Assisted Clear 2: Clear, 3: LIFE4, 4: Good FC (Full Combo), 5: Great FC, 6: PFC, 7: MFC */
export type ClearLamp = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7

/**
 * Request body to `/api/v1/scores/{:songId}/{:playStyle}/{:difficulty}`
 * @see https://github.com/ddradar/ddradar/blob/master/api/postChartScore/README.md
 */
export type Score = Pick<
  UserScore,
  'score' | 'exScore' | 'maxCombo' | 'clearLamp' | 'rank'
>

export function getDanceLevel(score: number) {
  const rankList = [
    { border: 990000, rank: 'AAA' },
    { border: 950000, rank: 'AA+' },
    { border: 900000, rank: 'AA' },
    { border: 890000, rank: 'AA-' },
    { border: 850000, rank: 'A+' },
    { border: 800000, rank: 'A' },
    { border: 790000, rank: 'A-' },
    { border: 750000, rank: 'B+' },
    { border: 700000, rank: 'B' },
    { border: 690000, rank: 'B-' },
    { border: 650000, rank: 'C+' },
    { border: 600000, rank: 'C' },
    { border: 590000, rank: 'C-' },
    { border: 550000, rank: 'D+' },
  ] as const
  for (const { border, rank } of rankList) {
    if (score >= border) return rank
  }
  return 'D'
}
