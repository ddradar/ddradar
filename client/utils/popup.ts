import type { BuefyNamespace } from 'buefy/types'

const option = { position: 'is-top', hasIcon: true } as const

export function success(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-success', ...option })
}

export function warning(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-warning', ...option })
}

export function danger(
  $buefy: Pick<BuefyNamespace, 'notification'>,
  message: string
) {
  $buefy.notification.open({ message, type: 'is-danger', ...option })
}
