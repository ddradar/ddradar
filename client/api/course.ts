import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

export function getCourseType(type: number) {
  return type === 1 ? 'NONSTOP' : type === 2 ? '段位認定' : ''
}

/**
 * Call "Get Course List" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/courses/
 */
export function getCourseList(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  series?: 16 | 17,
  type?: 1 | 2
) {
  const searchParams = new URLSearchParams()
  if (series) searchParams.append('series', `${series}`)
  if (type) searchParams.append('type', `${type}`)
  return $http.$get<Api.CourseListData[]>(`${apiPrefix}/courses`, {
    searchParams,
  })
}

/**
 * Call "Get Course Information" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/courses__id/
 */
export function getCourseInfo(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  id: string
) {
  return $http.$get<Api.CourseInfo>(`${apiPrefix}/courses/${id}`)
}
