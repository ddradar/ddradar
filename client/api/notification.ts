import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

/**
 * Call "Get Notification List" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification/
 */
export function getNotificationList(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  topOnly?: boolean
) {
  const query = topOnly ? '?scope=top' : ''
  return $http.$get<Api.NotificationListData[]>(
    `${apiPrefix}/notification${query}`
  )
}

/**
 * Call "Get Notification Info" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification__id/
 */
export function getNotificationInfo(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  id: string
) {
  return $http.$get<Api.NotificationInfo>(`${apiPrefix}/notification/${id}`)
}

/**
 * Call "Post Notification" API. (Admin only)
 * @see https://github.com/ddradar/ddradar/tree/master/api/notification--post/
 */
export function postNotification(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  body: Omit<Api.NotificationBody, 'sender'>
) {
  return $http.$post<Api.NotificationBody>(`${apiPrefix}/notification`, {
    ...body,
    sender: 'SYSTEM',
  })
}
