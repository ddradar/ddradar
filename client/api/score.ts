import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

/**
 * Call "Get Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--get/
 */
export function getChartScore(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4,
  scope?: 'private' | 'medium' | 'full'
) {
  const query = scope ? `?scope=${scope}` : ''
  return $http.$get<Api.ScoreInfo[]>(
    `${apiPrefix}/scores/${songId}/${playStyle}/${difficulty}${query}`
  )
}

/**
 * Call "Post Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--post/
 */
export function postChartScore(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4,
  score: Api.ScoreBody
) {
  return $http.$post(
    `${apiPrefix}/scores/${songId}/${playStyle}/${difficulty}`,
    score
  )
}

/**
 * Call "Delete Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--delete/
 */
export function deleteChartScore(
  $http: Pick<NuxtHTTPInstance, 'delete'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4
) {
  return $http.delete(`/api/v1/scores/${songId}/${playStyle}/${difficulty}`)
}

/**
 * Call "Post Song Scores" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id--post/
 */
export function postSongScores(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  songId: string,
  scores: Api.ScoreListBody[]
) {
  return $http.$post<Api.ScoreInfo[]>(`${apiPrefix}/scores/${songId}`, scores)
}
