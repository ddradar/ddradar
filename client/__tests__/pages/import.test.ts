import { scoreTexttoScoreList } from '@ddradar/core'
import {
  createLocalVue,
  mount,
  RouterLinkStub,
  shallowMount,
  Wrapper,
} from '@vue/test-utils'
import Buefy from 'buefy'
import { mocked } from 'ts-jest/utils'
import VueI18n from 'vue-i18n'

import { postSongScores } from '~/api/score'
import ImportPage from '~/pages/import.vue'
import * as popup from '~/utils/popup'

jest.mock('~/api/score')
jest.mock('~/utils/popup')
jest.mock('@ddradar/core')

const localVue = createLocalVue()
localVue.use(Buefy)
localVue.use(VueI18n)

describe('pages/import.vue', () => {
  let wrapper: Wrapper<ImportPage>
  const $accessor = { user: { code: 10000000 } }
  const $http = {}
  const $buefy = {}
  const i18n = new VueI18n({ locale: 'ja', silentFallbackWarn: true })
  const stubs = { NuxtLink: RouterLinkStub }
  beforeEach(() => {
    const mocks = { $accessor, $http, $buefy }
    wrapper = shallowMount(ImportPage, { localVue, mocks, i18n, stubs })
  })

  describe.each(['ja', 'en'])('{ locale: "%s" } snapshot test', locale => {
    const i18n = new VueI18n({ locale, silentFallbackWarn: true })
    test('renders correctly', () => {
      const data = () => ({ sourceCode: '<html></html>' })
      const mocks = { $accessor }
      const wrapper = mount(ImportPage, { localVue, data, mocks, i18n, stubs })
      expect(wrapper).toMatchSnapshot()
    })
    test('renders uploading state', () => {
      const wrapper = mount(ImportPage, {
        localVue,
        data: () => ({
          file: {},
          loading: true,
          maxCount: 20,
          doneCount: 7,
          currentSong: 'PARANOiA',
        }),
        mocks: { $accessor },
        i18n,
        stubs,
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  // Computed
  describe('get bookmarklet()', () => {
    test.each([
      ['en', /.+\/eagate\.en\.min\.js/],
      ['ja', /.+\/eagate\.ja\.min\.js/],
    ])('{ locale: %s } returns %s script', (locale, expected) => {
      const i18n = new VueI18n({ locale, silentFallbackWarn: true })
      const mocks = { $accessor }
      const wrapper = shallowMount(ImportPage, { localVue, mocks, i18n, stubs })

      // Act - Assert
      // @ts-ignore
      expect(wrapper.vm.bookmarklet).toMatch(expected)
    })
  })

  // Method
  describe('copyToClipboard()', () => {
    Object.assign(navigator, { clipboard: { writeText: () => {} } })
    jest.spyOn(navigator.clipboard, 'writeText')
    test('copies bookmarklet text to clipboard', async () => {
      const mocks = { $accessor }
      const wrapper = shallowMount(ImportPage, { localVue, mocks, i18n, stubs })

      // @ts-ignore
      const expected = wrapper.vm.bookmarklet
      // @ts-ignore
      await wrapper.vm.copyToClipboard()

      expect(navigator.clipboard.writeText).toBeCalledWith(expected)
      expect(mocked(popup.success)).toBeCalled()
    })
  })
  describe('importSkillAttackScores()', () => {
    const scoreData = [
      {
        songName: '朧',
        playStyle: 1,
        difficulty: 0,
        score: 876000,
        clearLamp: 2,
        rank: 'A+',
      } as const,
      {
        songName: '朧',
        playStyle: 1,
        difficulty: 2,
        score: 823000,
        clearLamp: 2,
        rank: 'A',
      } as const,
      {
        songName: '朧',
        playStyle: 1,
        difficulty: 3,
        score: 798000,
        clearLamp: 2,
        rank: 'A-',
      } as const,
      {
        songName: '朧',
        playStyle: 1,
        difficulty: 4,
        score: 780000,
        clearLamp: 2,
        rank: 'B+',
      } as const,
    ]
    const scoreList: ReturnType<typeof scoreTexttoScoreList> = {
      '691': scoreData,
    }
    const postMock = mocked(postSongScores)
    beforeEach(() => {
      postMock.mockClear()
      mocked(scoreTexttoScoreList).mockClear()
      mocked(popup.success).mockClear()
      mocked(popup.warning).mockClear()
      mocked(popup.danger).mockClear()
    })

    test('calls "Post Song Scores" API', async () => {
      // Arrange
      wrapper.setData({ file: {} })
      await wrapper.vm.$nextTick()
      mocked(scoreTexttoScoreList).mockReturnValue(scoreList)

      // Act
      // @ts-ignore
      await wrapper.vm.importSkillAttackScores()

      // Assert
      expect(postMock).toBeCalledWith($http, '691', scoreList['691'])
      expect(mocked(popup.success)).toBeCalled()
    })
    test('does not call scoreTexttoScoreList() if file is empty', async () => {
      // Arrange
      wrapper.setData({ file: null })
      await wrapper.vm.$nextTick()
      mocked(scoreTexttoScoreList).mockReturnValue({ foo: [] })

      // Act
      // @ts-ignore
      await wrapper.vm.importSkillAttackScores()

      // Assert
      expect(mocked(scoreTexttoScoreList)).not.toBeCalled()
      expect(postMock).not.toBeCalled()
    })
    test('does not call "Post Song Scores" API if scores is empty', async () => {
      // Arrange
      wrapper.setData({ file: {} })
      await wrapper.vm.$nextTick()
      mocked(scoreTexttoScoreList).mockReturnValue({ foo: [] })

      // Act
      // @ts-ignore
      await wrapper.vm.importSkillAttackScores()

      // Assert
      expect(mocked(scoreTexttoScoreList)).toBeCalled()
      expect(mocked(popup.success)).toBeCalled()
      expect(postMock).not.toBeCalled()
    })
    test.each([
      ['ja', 'ファイルが読み取れないか、正しい形式ではありません'],
      ['en', 'Cannot read file or invalid file'],
    ])(
      '{ locale: "%s" } warns "%s" if sourceCode is invalid',
      async (locale, message) => {
        // Arrange
        mocked(scoreTexttoScoreList).mockImplementation(() => {
          throw new Error('invalid')
        })
        const wrapper = shallowMount(ImportPage, {
          localVue,
          mocks: { $accessor, $http, $buefy },
          i18n: new VueI18n({ locale, silentFallbackWarn: true }),
          data: () => ({ file: {} }),
          stubs,
        })

        // Act
        // @ts-ignore
        await wrapper.vm.importSkillAttackScores()

        // Assert
        expect(mocked(popup.warning)).toBeCalledWith($buefy, message)
      }
    )
    test('calls popup.danger() if "Post Song Scores" API throws error', async () => {
      // Arrange
      wrapper.setData({ file: {} })
      await wrapper.vm.$nextTick()
      mocked(scoreTexttoScoreList).mockReturnValue(scoreList)
      postMock.mockRejectedValue('Error')

      // Act
      // @ts-ignore
      await wrapper.vm.importSkillAttackScores()

      // Assert
      expect(postMock).toBeCalledWith($http, '691', scoreList['691'])
      expect(mocked(popup.success)).not.toBeCalled()
      expect(mocked(popup.danger)).toBeCalledWith($buefy, 'Error')
    })
  })
})
