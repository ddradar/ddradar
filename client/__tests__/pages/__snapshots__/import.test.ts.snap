// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/import.vue renders correctly 1`] = `
<section class="section">
  <h1 class="title">スコアのインポート</h1>
  <h2 class="subtitle">e-amusementサイトからインポート</h2>
  <div class="content">
    <p>
      公式サイトの「楽曲データ一覧」ページより、スコアを取得します。<wbr> <strong>
        公式サイトでのスコア閲覧には、ベーシックコースへの加入が必要です。
      </strong></p>
    <div class="field"><label class="label">HTMLソース</label>
      <div class="control is-clearfix"><textarea required="required" class="textarea"></textarea>
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
    <div class="field">
      <!----> <button type="button" disabled="disabled" class="button is-success">
        <!----> <span>
        登録
      </span>
        <!----></button>
      <!---->
    </div>
  </div>
</section>
`;
